storey/timescaledb_target.py

```diff
@@ -0,0 +1,322 @@
+        ...     columns=["$key", "value", "source=device_name"],
+        ...     table="events"
+        ... )
+
+    Note:
+        - The target table must be created as a TimescaleDB hypertable before use
+        - The time column must be of TIMESTAMPTZ type for proper time-series functionality
+        - Events are written using PostgreSQL's COPY protocol for optimal performance
+        - Connection pooling is handled automatically with proper cleanup on termination
+    """
+
+    def __init__(
+        self,
+        dsn: str,
+        time_col: str,
+        columns: list[str],
+        time_format: Optional[str] = None,
+        table: Optional[str] = None,
+        max_connections: int = 10,
+        min_connections: int = 10,
+        **kwargs,
+    ) -> None:
+
+        # Pass parameters to parent classes
+        kwargs["time_col"] = time_col
+        kwargs["columns"] = columns
+        if table:
+            kwargs["table"] = table
+        if time_format:
+            kwargs["time_format"] = time_format
+
+        # Initialize batching functionality
+        _Batching.__init__(self, **kwargs)
+
+        # Initialize writer functionality with time column as first column
+        # This ensures proper column ordering for TimescaleDB's time-partitioned structure
+        _Writer.__init__(
+            self,
+            columns=[time_col] + columns,
+            infer_columns_from_data=False,
+            retain_dict=True,
+            time_field=time_col,
+            time_format=time_format,
+        )
+        self._table = table
+
+        # Store configuration
+        self._time_col = time_col
+        self._columns = columns
+
+        # Database connection configuration
+        self._dsn = dsn
+        self._max_connections = max_connections
+        self._min_connections = min_connections
+        self._pool = None  # Connection pool will be created lazily during first use
+
+    def _test_connection_sync(self) -> None:
+        """Test database connection synchronously during initialization.
+
+        This method validates that the TimescaleDB instance is accessible and properly configured
+        before any events are processed. It performs the following checks:
+        1. Basic connectivity to PostgreSQL/TimescaleDB
+        2. Table existence validation (if table name is provided)
+        3. TimescaleDB extension availability (optional check)
+
+        Raises:
+            ConnectionError: If connection fails, table doesn't exist, or other database issues occur
+        """
+        try:
+            # Get or create event loop for sync context
+            try:
+                loop = asyncio.get_event_loop()
+            except RuntimeError:
+                # No event loop exists, create one
+                loop = asyncio.new_event_loop()
+                asyncio.set_event_loop(loop)
+
+            # Test connection synchronously
+            loop.run_until_complete(self._test_connection_async())
```

Comment:
I'm pretty sure this will fail with `RuntimeError: This event loop is already running
` if actually run from an async context (e.g. nuclio, or with an `AsyncEmitSource`).